### ISSUE TITLE
Support efficient pointer arithmetic with non-power-of-2 structure sizes

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#85]
Fix this obscure problem with pointer arithmetic on non-power-of-2 structure addresses: snabbco/snabb#1207.

**Comment by lukego** [comment]
Here is a program that fails to JIT a non-power-of-2 structure size (tested with LuaJIT):

```lua
local ct = require("ffi").typeof("struct { char x[12]; } *")
local x
for i = 1, 1000 do
   x = ct() - ct()
end
```

Giving:

```
---- TRACE 1 start test.lua:3
0012  MOV      6   0
0013  CALL     6   2   1
0000  . FUNCC               ; ffi.meta.__call
0014  MOV      7   0
0015  CALL     7   2   1
0000  . FUNCC               ; ffi.meta.__call
0016  SUBVV    1   6   7
0000  . . FUNCC               ; ffi.meta.__sub
---- TRACE 1 abort test.lua:4 -- bad argument type
```

and the relevant line (in raptorjit) seems to be here: https://github.com/raptorjit/raptorjit/blob/master/src/lj_crecord.c#L1247

**Comment by lukego** [comment]
Is this a case of "the best is the enemy of the good"?

Seems like a "good" solution would be to simply use division (`IDIV` on x86) and pay the high-but-not-ridiculous price of that (~22 cycles latency for 64-bit.) However, perhaps no self-respecting compiler hacker would ever use a hardware division instruction to divide by a constant when better methods are known (https://github.com/snabbco/snabb/issues/1207#issuecomment-321867454), and so it was simply left as an NYI until time would permit to implement the "best" solution?

I am not a self-respecting compiler hacker and so I have no qualms about proposing this draft "short term" change to fall back on plain ol' division to keep the code on-trace in JIT mode:

```diff
diff --git a/src/lj_crecord.c b/src/lj_crecord.c
index 2db4c86..afdb15d 100644
--- a/src/lj_crecord.c
+++ b/src/lj_crecord.c
@@ -1314,14 +1314,19 @@ static TRef crec_arith_ptr(jit_State *J, TRef *sp, CType **s, MMS mm)
       if (mm == MM_sub) {  /* Pointer difference. */
        TRef tr;
        CTSize sz = lj_ctype_size(cts, ctype_cid(ctp->info));
-       if (sz == 0 || (sz & (sz-1)) != 0)
-         return 0;  /* NYI: integer division. */
-       tr = emitir(IRT(IR_SUB, IRT_INTP), sp[0], sp[1]);
-       tr = emitir(IRT(IR_BSAR, IRT_INTP), tr, lj_ir_kint(J, lj_fls(sz)));
+       if (sz == 0) {
+          return 0;
+        } else if ((sz & (sz-1)) != 0) {
+          tr = emitir(IRT(IR_SUB, IRT_INTP), sp[0], sp[1]);
+          tr = emitir(IRT(IR_DIV, IRT_INTP), tr, lj_ir_kint(J, sz));
+        } else {
+          tr = emitir(IRT(IR_SUB, IRT_INTP), sp[0], sp[1]);
+          tr = emitir(IRT(IR_BSAR, IRT_INTP), tr, lj_ir_kint(J, lj_fls(sz)));
+        }
 #if LJ_64
-       tr = emitconv(tr, IRT_NUM, IRT_INTP, 0);
+        tr = emitconv(tr, IRT_NUM, IRT_INTP, 0);
 #endif
-       return tr;
+        return tr;
       } else {  /* Pointer comparison (unsigned). */
        /* Assume true comparison. Fixup and emit pending guard later. */
        IROp op = mm == MM_eq ? IR_EQ : mm == MM_lt ? IR_ULT : IR_ULE;
```

cc @alexandergall maybe you want to test that patch?

aside: I did a little microbenchmarking to see if Intel CPUs were already recognizing some small division operands and optimizing them specially, but I don't think so, e.g. I do see ~22 cycles latency to divide by 12. The CPU designers may see this kind of optimization as the domain of compilers.

**Comment by lukego** [comment]
Relatedly: [Code](https://github.com/raptorjit/raptorjit/blob/master/src/lj_opt_fold.c#L955-L983) for "constant folding optimization" handling the case where a number is divided by a constant. It handles some simple cases like `x / 1 <=> x` and `x * 2 <-> x + x` but not `x / k`. Could optimization be done here in a more general way rather than inside the trace recorder on pointer-difference?

**Comment by lukego** [comment]
Interesting. If I am understanding the x86 [IDIV](http://www.tptp.cc/mirrors/siyobik.info/instruction/IDIV.html) instruction correctly then there are two variants that we could choose here: to divide a 128-bit numerator by a 64-bit divisor (e.g. `div rbx` operating on `rdx:rax`), or to divide a 64-bit numerator by a 32-bit divisor (e.g. `div ebx` operating on `edx:eax`). The performance of each variant is quite different.

- 128-by-64 division has 42-95 cycles latency (quoth Agner Fog) on Skylake. That is similar to between one or two L3 cache access (42 cycles).
- 64-by-32 division has 26 cycles latency on Skylake. That is similar to two L2 cache accesses (12 cycles) or ~2/3 of an L3 cache access.

So what code do we get with my patch above to use a `DIV` IR instruction? In fact the mcode will compile to a call to the C function [`lj_carith_divi64`](https://github.com/raptorjit/raptorjit/blob/master/src/lj_carith.c#L347) that in turn compiles like this:

```
0000000000453120 <lj_carith_divi64>:
  453120:       48 b8 00 00 00 00 00    movabs rax,0x8000000000000000
  453127:       00 00 80
  45312a:       48 85 f6                test   rsi,rsi
  45312d:       74 13                   je     453142 <lj_carith_divi64+0x22>
  45312f:       48 39 c7                cmp    rdi,rax
  453132:       75 06                   jne    45313a <lj_carith_divi64+0x1a>
  453134:       48 83 fe ff             cmp    rsi,0xffffffffffffffff
  453138:       74 08                   je     453142 <lj_carith_divi64+0x22>
  45313a:       48 89 f8                mov    rax,rdi
  45313d:       48 99                   cqo
  45313f:       48 f7 fe                idiv   rsi
  453142:       c3                      ret
  453143:       66 66 66 66 2e 0f 1f    data16 data16 data16 nop WORD PTR cs:[rax+rax*1+0x0]
  45314a:       84 00 00 00 00 00
```

So this compiles to the expensive 128-by-64 division instruction (`idiv rsi`) in addition to the overhead of the C function call and some guards in the C code. This is probably quite expensive.

It may well be worth special casing the C code for when the divisor is only 32-bit (e.g. size of a C struct is seldom more than 4GB), although if we were to accumulate too many special cases then it may be worth skipping straight to the "Right Thing" of generating special code for constant divisors that doesn't do any hardware division.

**Comment by alexandergall** [comment]
I have verified that the trace aborts disappear whit this patch. I also see `lj_carith_divi64` show up in `perf top` but pretty far down the list (at a little under 0.3%).

**Comment by lukego** [comment]
@alexandergall Groovy. I'm assuming that with this patch the overall performance you see is much better than with the trace aborts, closer to the power-of-2 version?

**Comment by alexandergall** [comment]
Yes, I think so. Currently, the IPv6 reassembly code, which is my test-case for this, has other performance issues in context of the `l2vpn` program, which makes it hard to quantify these effects.

**Comment by lukego** [comment]
This idea of mine does not have legs:

> It may well be worth special casing the C code for when the divisor is only 32-bit

Because even if a 32-bit divisor would be okay it would not be cool to have the result (quotient) truncated to 32-bit. So we need to use the full 64-bit divide instruction even if we know the divisor is small.

I had never before considered that `&a - &b` pointer difference is in principle much more expensive on 64-bit machines than 32-bit because the semantics requires a division. I suppose it makes sense to use this operation sparingly when its exact semantics are actually what you need.
